crates/rrg/src/action/yara_scan_memory.rs

```diff
@@ -0,0 +1,564 @@
+// Additionally, we want to send matching data to the blob store, instead of sending it back
+// inline in responses.
+
+#[derive(Debug)]
+struct Rule {
+    identifier: String,
+    patterns: Vec<Pattern>,
+}
+
+#[derive(Debug)]
+struct Pattern {
+    identifier: String,
+    matches: Vec<Match>,
+}
+
+#[derive(Debug)]
+struct Match {
+    /// Address range in the process' address space at which the match was found.
+    range: std::ops::Range<usize>,
+    /// Sha256 digest of the matching bytes that were sent to the blob sink.
+    blob_sha256: [u8; 32],
+}
+
+impl Rule {
+    /// Converts a [yara_x::Rule] to a [Rule], sending any matching data to the blob store
+    /// and storing the blob's Sha256 hash in the corresponding [Match].
+    fn blobify<S: crate::session::Session>(
+        rule: yara_x::Rule<'_, '_>,
+        session: &mut S,
+    ) -> crate::session::Result<Self> {
+        use sha2::Digest as _;
+        let patterns = rule
+            .patterns()
+            .map(|pattern| {
+                let matches = pattern
+                    .matches()
+                    .map(|mat| {
+                        let blob_sha256 = sha2::Sha256::digest(mat.data()).into();
+                        let blob = crate::blob::Blob::from(mat.data().to_vec());
+                        session.send(crate::Sink::Blob, blob)?;
+                        Ok(Match {
+                            range: mat.range(),
+                            blob_sha256,
+                        })
+                    })
+                    .collect::<crate::session::Result<_>>()?;
+                Ok(Pattern {
+                    identifier: pattern.identifier().to_owned(),
+                    matches,
+                })
+            })
+            .collect::<crate::session::Result<_>>()?;
+        Ok(Self {
+            identifier: rule.identifier().to_owned(),
+            patterns,
+        })
+    }
+}
+
+impl From<Rule> for rrg_proto::yara_scan_memory::Rule {
+    fn from(rule: Rule) -> Self {
+        use rrg_proto::yara_scan_memory as proto;
+
+        let mut ret = Self::new();
+        ret.set_identifier(rule.identifier);
+        ret.set_patterns(
+            rule.patterns
+                .into_iter()
+                .map(|pattern| {
+                    let mut ret = proto::Pattern::new();
+                    ret.set_identifier(pattern.identifier);
+                    ret.set_matches(
+                        pattern
+                            .matches
+                            .into_iter()
+                            .map(|mat| {
+                                let mut ret = proto::Match::new();
+                                ret.set_offset(mat.range.start as u64);
+                                ret.set_data_sha256(mat.blob_sha256.to_vec());
+                                ret
+                            })
+                            .collect(),
+                    );
+                    ret
+                })
```

Comment:
Why not implement `From` for each corresponding proto? In contrast to my previous post, these seem like excellent abstraction points.